### PR TITLE
Fuzz base64/v2

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6050

### CODE REVIEW

**Comment by catenacyber** [comment]
@inashivb did you investigate the buffer overflow found by CIFuzz with this new target ?

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        uint32_t dest_size = 3 * len / 4;
+        uint8_t *dest = malloc(dest_size);
+        uint32_t consumed_bytes = 0;
+        uint32_t decoded_bytes = 0;
+
+        DecodeBase64(dest, dest_size, src, len, &consumed_bytes, &decoded_bytes, mode);
+    }
+}
+
+static void AmpleDestinationSizeTest(const uint8_t *src, size_t len)
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:27]: Why do we need `AmpleDestinationSizeTest` ? Is `ConstrainedDestinationSizeTest` not enough ?

File: src/tests/fuzz/fuzz_decodebase64.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include "util-base64.h"
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+static int initialized = 0;
+
+static void ConstrainedDestinationSizeTest(const uint8_t *src, size_t len)
+{
+    for (uint8_t mode = BASE64_MODE_RELAX; mode <= BASE64_MODE_RFC4648; mode++) {
+        uint32_t dest_size = 3 * len / 4;
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_decodebase64.c:18]: I think it is `3 * ((len + 3)/ 4)`

Check with len = 1 ;-)

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 14281

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 2 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:18]: Indeed. Missing a very apparent edge case.

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 1 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: I thought that if due to constrained size if some of the data wasn't fully decoded but was capable of hitting a Debug validation then this would help. Wdyt?

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 2 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:18]: Wait. We want to constrain the space here. I think I kept this here this way because dest_size is controlled by the callers. So, what if the caller has run out of space in the dest buffer but still has a byte to decode? Would it not still make a valid test case?

**Comment by inashivb** [comment]
> @inashivb did you investigate the buffer overflow found by CIFuzz with this new target ?

Just did. Seems to be on the case you pointed out. 0 bytes on destination buffer and length of the source string 1.
Please let me know what you think of what I think in the inline comments.

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 1 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: One example is the case you pointed out where we don't have enough dest space for decoding 1byte. While that should be checked as that may happen, it should also be checked that if there is just 1 byte and enough destination space, it actually does get decoded properly.

Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: But `ConstrainedDestinationSizeTest` should always have enough space, right ?

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 2 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:18]: Does Suricata want to partially decode base64 ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
Will you do a new iteration Shivani ?

**Comment by inashivb** [comment]
> Will you do a new iteration Shivani ?

yes. Philippe. I also have to answer your questions. I'm first trying to do #9160 as it is for the release. While doing that, I'm also able to find answers to some questions.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 14281

**Comment by inashivb** [comment]
Needs more work. Is stale and has conflicts now.

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 1 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: destination size is controlled by the caller and may not have enough space, I think.
refs:
1. https://github.com/OISF/suricata/blob/master/src/util-decode-mime.c#L1225
2. https://github.com/OISF/suricata/blob/master/src/util-base64.c#L178

Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: wdyt?

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 2 above)
Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:18]: yes, in case we run out of space to decode more data, I think. wdyt?

Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:18]: So, `dest_size` should be random from the input ?

File: src/tests/fuzz/fuzz_decodebase64.c (hunk 1 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: I guess I would go for only one test, where the destination size is fuzzingly randomized

Comment (inashivb) [reply src/tests/fuzz/fuzz_decodebase64.c:27]: That sounds a lot better. Thank you!
